templates/cargo_crate.gni

```
@@ -0,0 +1,193 @@
+    _rustenv += [ "CARGO_PKG_VERSION_MAJOR=${_cargo_pkg_version_parts[0]}" ]
+    _rustenv += [ "CARGO_PKG_VERSION_MINOR=${_cargo_pkg_version_parts[1]}" ]
+    _cargo_pkg_version_subparts = string_split(_cargo_pkg_version_parts[2], "-")
+    _cargo_pkg_version_subparts += [ "" ]  # In case there is no prerelease.
+    _rustenv += [ "CARGO_PKG_VERSION_PATCH=${_cargo_pkg_version_subparts[0]}" ]
+    _rustenv += [ "CARGO_PKG_VERSION_PRE=${_cargo_pkg_version_subparts[1]}" ]
+  }
+  if (defined(invoker.cargo_pkg_name)) {
+    _rustenv += [ "CARGO_PKG_NAME=${invoker.cargo_pkg_name}" ]
+  }
+  if (defined(invoker.cargo_pkg_repository)) {
+    _rustenv += [ "CARGO_PKG_REPOSITORY=${invoker.cargo_pkg_repository}" ]
+  }
+
+  # Try to determine the CARGO_MANIFEST_DIR, preferring the directory
+  # with build.rs and otherwise assuming that the target contains a
+  # `crate/` subdirectory.
+  if (defined(invoker.build_root)) {
+    manifest_dir = "."
+    _rustenv += [ "CARGO_MANIFEST_DIR=${manifest_dir}" ]
+  }
+
+  build_rust(target_name) {
+    forward_variables_from(invoker,
+                           "*",
+                           [
+                             "build_root",
+                             "build_sources",
+                             "build_script_inputs",
+                             "build_script_outputs",
+                             "rustenv",
+                           ])
+    rustenv = _rustenv
+    if (!defined(rustflags)) {
+      rustflags = []
+    }
+
+    crate_type = "rlib"
+    if (defined(invoker.crate_type)) {
+      crate_type = invoker.crate_type
+    }
+
+    if (defined(invoker.build_root)) {
+      if (!defined(deps)) {
+        deps = []
+      }
+      if (!defined(sources)) {
+        sources = []
+      }
+
+      flags_file = "$target_out_dir/cargo_flags.rs"
+
+      rustflags += [ "@" + rebase_path(flags_file, root_build_dir) ]
+      sources += [ flags_file ]
+      if (defined(invoker.build_script_outputs)) {
+        inputs = []
+        foreach(extra_source,
+                filter_exclude(invoker.build_script_outputs, [ "*.rs" ])) {
+          inputs += [ "$target_out_dir/$extra_source" ]
+        }
+        foreach(extra_source,
+                filter_include(invoker.build_script_outputs, [ "*.rs" ])) {
+          sources += [ "$target_out_dir/$extra_source" ]
+        }
+        _out_dir = rebase_path(target_out_dir)
+        rustenv += [ "OUT_DIR=$_out_dir" ]
+      }
+      deps += [ ":${build_script_name}_output" ]
+    }
+  }
+
+  if (defined(invoker.build_root)) {
+    action("${build_script_name}_output") {
+      script = rebase_path("//build/scripts/run_build_script.py")
+      build_script_target = ":${build_script_name}(//build/toolchain:host)"
+      deps = [ build_script_target ]
+      build_script = "$root_build_dir/host/bin/${build_script_name}"
+      flags_file = "$target_out_dir/cargo_flags.rs"
+      args = [
+        "--build-script",
```

```suggestion
        "--build_script",
```
Let's use google-style command line flags.